check for uninitialized fields when closing the Transport

kelmenhorst opened the pull request:
The underlying observation that motivates this patch is that `(*Transport.)Close` and `(*Transport.)close` try to access possibly uninitialized fields of the `Transport` which can lead to nil pointer dereference failures (#3906) when the Transport's `init` function is not (fully) executed.

To fix this, I added respective checks in `(*Transport.)Close` and `(*Transport.)close`, so that the robustness of calling `(*Transport.)Close` does not depend on whether or how much of `(Transport.)init` was executed before. 
However, it might be that the underlying issue needs to be addressed at another place (I am not familiar enough with the code base to fully evaluate this).

marten-seemann reviewed (commented):
1. Can you please sign the CLA?
2. A test case would be appreciated.

kelmenhorst commented:
Sorry, I had forgot to refresh the CLA check after signing it.
I added a test case to transport_test.go.

Apart from the nil-checks, I changed `(*Transport.)Close` such that it closes the underlying `Conn` (net.PacketConn), rather than `conn` (rawConn), since `conn` might not be initialized.

Btw, what's the reason for allowing uninitialized `Transport`'s anyway?

marten-seemann commented:
> Btw, what's the reason for allowing uninitialized Transport's anyway?

What do you mean? It's an exported struct, so users can just use it.

marten-seemann commented:
> Btw, what's the reason for allowing uninitialized Transport's anyway?

What do you mean? It's an exported struct, so users can just use it.

marten-seemann reviewed (approved):
Thank you @kelmenhorst!

kelmenhorst commented:
> What do you mean? It's an exported struct, so users can just use it.

Yes, I guess I was asking whether it would make sense to not export the struct itself, and rather export a constructor that initializes the Transport.
However, I can imagine that there are use cases that I am not aware of where you need more flexibility with the Transport :)

marten-seemann commented:
I see the point, but it's nice to have some configuration fields on the `Transport`. And it matches the standard library pattern (e.g. `net.Dialer`).
